tinker_cookbook/supervised/train.py

```diff
@@ -227,30 +166,114 @@ def main(config: Config):
+        metrics["progress"] = step / progress_denominator
+
+        learning_rate = config.learning_rate * compute_schedule_lr_multiplier(
+            lr_schedule=config.lr_schedule,
+            step=step,
+            total_steps=total_steps,
+        )
+        metrics["learning_rate"] = learning_rate
+
+        adam_params = tinker.AdamParams(
+            learning_rate=learning_rate,
+            beta1=config.adam_beta1,
+            beta2=config.adam_beta2,
+            eps=config.adam_eps,
+        )
+
+        with timed("get_batch", metrics):
+            data = dataset.get_batch(batch_idx)
+        if data:
+            logger.info(colorize_example(data[0], tokenizer))
+
+        fwd_bwd_future = await training_client.forward_backward_async(data, loss_fn="cross_entropy")
+        optim_step_future = await training_client.optim_step_async(adam_params)
+
+        return SubmittedBatch(
+            fwd_bwd_future=fwd_bwd_future,
+            optim_step_future=optim_step_future,
+            metrics=metrics,
+            data=data,
+            step=step,
+            epoch_idx=epoch_idx,
+            batch_idx=batch_idx,
+            batch_start_time=batch_start_time,
+        )
+
+    async def finish_batch(submitted: SubmittedBatch):
+        metrics = submitted.metrics
+        metrics["progress"] = min((submitted.step + 1) / progress_denominator, 1.0)
+
+        if submitted.step % config.save_every == 0 and submitted.step > 0:
+            with timed("save_checkpoint", metrics):
+                checkpoint_paths = await checkpoint_utils.save_checkpoint_async(
+                    training_client=training_client,
+                    name=f"{submitted.step:06d}",
+                    log_path=config.log_path,
+                    loop_state={"epoch": submitted.epoch_idx, "batch": submitted.batch_idx},
+                    kind="both",
+                )
+            metrics.update(checkpoint_paths)
+
+        with timed("step", metrics):
+            fwd_bwd_result = await submitted.fwd_bwd_future.result_async()
+            await submitted.optim_step_future.result_async()
+
+        logprobs = [x["logprobs"] for x in fwd_bwd_result.loss_fn_outputs]
+        weights = [datum.loss_fn_inputs["weights"] for datum in submitted.data]
+        train_nll = compute_mean_nll(logprobs, weights)
+
+        metrics.update(
+            num_sequences=len(submitted.data),
+            num_tokens=sum(datum.model_input.length for datum in submitted.data),
+            num_loss_tokens=sum(
+                sum(datum.loss_fn_inputs["weights"].data) for datum in submitted.data
+            ),
+            train_mean_nll=train_nll,
+        )
+        metrics["time/total"] = time.time() - submitted.batch_start_time
+
+        if evaluators and config.eval_every > 0 and submitted.step % config.eval_every == 0:
+            with timed("evals", metrics):
+                eval_metrics = await run_evals(evaluators, training_client, submitted.step)
```

Comment:
If running evals synchronously on the main training loop is a bottleneck in the future, we could run the evals on a background coroutine -- this would allow overlapping the next step to train while we do evals on the current step. The downside here is that we won't be able to log the eval metrics in the same event as the normal training step metrics (but that might be OK from wandb perspective).

Reply:
yeah that's a good idea. wandb might already handle it correctly. the jsonl loading would need to do something slightly more complicated, but it's not too bad.

Reply:
i'd be in favor of that -- @rafailovtm needed and implemented this previously, but at the time it caused too much complexity to merge.